compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -812,17 +814,103 @@ trait Implicits { self: Typer =>
           EmptyTree
       }
 
+  lazy val synthesizedProductMirror: SpecialHandler =
+    (formal: Type, span: Span) => implicit (ctx: Context) => {
+      formal.member(tpnme.MonoType).info match {
+        case monoAlias @ TypeAlias(monoType) =>
+          if (monoType.termSymbol.is(CaseVal)) {
+            val modul = monoType.termSymbol
+            val caseLabel = ConstantType(Constant(modul.name.toString))
+            val mirrorType = defn.Mirror_SingletonType
+              .refinedWith(tpnme.MonoType, monoAlias)
+              .refinedWith(tpnme.CaseLabel, TypeAlias(caseLabel))
+            ref(modul).withSpan(span).cast(mirrorType)
+          }
+          else if (monoType.classSymbol.isGenericProduct) {
+            val cls = monoType.classSymbol
+            val accessors = cls.caseAccessors.filterNot(_.is(PrivateLocal))
+            val elemTypes = accessors.map(monoType.memberInfo(_))
+            val caseLabel = ConstantType(Constant(cls.name.toString))
+            val elemLabels = accessors.map(acc => ConstantType(Constant(acc.name.toString)))
+            val mirrorType =
+              defn.Mirror_ProductType
+                .refinedWith(tpnme.MonoType, monoAlias)
+                .refinedWith(tpnme.ElemTypes, TypeAlias(TypeOps.nestedPairs(elemTypes)))
+                .refinedWith(tpnme.CaseLabel, TypeAlias(caseLabel))
+                .refinedWith(tpnme.ElemLabels, TypeAlias(TypeOps.nestedPairs(elemLabels)))
+            val modul = cls.linkedClass.sourceModule
+            assert(modul.is(Module))
+            ref(modul).withSpan(span).cast(mirrorType)
```

Comment:
What's preventing us from simply using actual type of `modul` here?

Reply:
I a bit confused about these implicits. If we need them, why synthesize them as opposed to adding an `implicit val mirror: ... = this` in companion objects?

Reply:
Because there is not always a companion object

 - sealed traits might come without one
 - enum cases don't have one either